doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,432 @@
+        ray.get(self._generation_signal.disable_generation.remote())
+        try:
+            yield
+        finally:
+            # Always re-enable to avoid deadlocking the driver on failures.
+            ray.get(self._generation_signal.allow_generation.remote())
+
+    @ray.method(tensor_transport="nixl")
+    def generate(self, states: torch.Tensor):
+        ray.get(self._generation_signal.wait_for_generation.remote())
+        with torch.no_grad():
+            states_cuda = states.cuda()
+            logits = self.model(states_cuda)  # [batch_size, ACTION_DIM]
+
+            # GRPO requires sampling from the current policy (not just the greedy action)
+
+            # Create distribution for each state (batch_size distributions over ACTION_DIM actions each)
+            dist = Categorical(logits=logits)
+            # Sample ACTION_DIM actions from each state's distribution.
+            acts = dist.sample((ACTION_DIM,))  # [ACTION_DIM, batch_size]
+            # Vectorized log_prob over the sampled actions for each state
+            logps = dist.log_prob(acts)  # [ACTION_DIM, batch_size]
+            acts = acts.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+            logps = logps.transpose(0, 1).contiguous()  # [batch_size, ACTION_DIM]
+
+        # Create trajectory slices and enqueue trajectory slices for Scorer
+        slice_batch = {
+            "policy_version": self.policy_version,
+            "state": states,
+            "actions": acts,
+            "old_logps": logps,
+        }
+        self.scorer.enqueue_trajectory_batch.remote(slice_batch)
+
+    def update_weights(self, cuda_weights, version: int) -> bool:
+        """Apply GPU-to-GPU policy weight updates while pausing sampling."""
+        # Disable sampling while the GPU object transfer and weight load are in-flight.
+        with self._pause_generation():
+            first_tensor = next(iter(cuda_weights.values()))
+            if first_tensor.device.type != "cuda":
+                raise RuntimeError(
+                    "Expected CUDA tensors after GPU-to-GPU direct transfer"
+                )
+            self.model.load_state_dict(cuda_weights)
+            self.model.eval()
+            self.policy_version = int(version)
+        return True
+
+
+def run_once(total_steps: int) -> None:
+    """Run one end-to-end training session."""
+    replay_buf = ReplayBuffer.remote()
+    learner = Learner.remote(replay_buf)
+    scorer = Scorer.remote(replay_buf)
+    signal = SignalActor.remote()
+    generator = Generator.remote(scorer, signal)
+
+    # Initialize generator with current learner weights
+    ray.get(
+        generator.update_weights.remote(
+            learner.get_weights_ref.remote(), learner.get_version.remote()
+        )
+    )
+
+    num_steps = total_steps
+
+    # Pre-fill ReplayBuffer before starting PPO.
+    ray.get(generator.generate.remote(sample_unit_vector(batch_size=BATCH_SIZE)))
```

Comment:
I believe this only waits for the generator.generate task to finish, but there may not be data in the replay buffer yet.

We can make this more robust by allowing ReplayBuffer.sample to return an empty batch.